File: lua/css.lua
-- css.lua
--
-- simple openresty handler to generate adblock-friendly CSS
--
-- calls a remote ad service, parses through your existing Adblock Lua code, returns in CSS template
--
-- nginx configuration
--
-- location "^/css/(.+)/(.+)$" {
--     set $ad_zone $1;
--     set $css_callback $2
--     lua_code_cache off; # fast reloading but less performant
--     content_by_lua_file /path/to/css.lua;
--     expires -1;
--     default_type text/css;
-- }
--
function main()
    -- tell adblock.lua to encode images/videos as base64
    package.path = package.path .. ';/usr/local/openresty/nginx/conf/?.lua'
    ngx.ctx['encode_video'] = 1
    ngx.ctx['encode_images'] = 1
    ngx.ctx['output_format'] = 1
    local css_template = '#%s{content:"%s"}'
    cjson = require "cjson"
    require "adblock"
    local ad_zone = ngx.var.ad_zone or 32
    local css_callback = ngx.var.css_callback or "x_000000"
    local client_ip = ngx.var.remote_addr
    local ad_url = string.format( "http://127.0.0.1/ads/%s?format=json&clientip=%s", ad_zone, client_ip)
    --ngx.log(ngx.ERR, "calling URL: ", ad_url)
    -- get the data from the ad provider
    local body, headers = fetch_url( ad_url )
    --ngx.log(ngx.ERR, "response body: ", body)
    -- process the ad: encode images/video as base64, massage domain names to be adblock friendly, etc
    local json_ad = process_ads_response( body, headers, ad_zone )
    --ngx.log(ngx.ERR, "json_ad: ", json_ad)
    local out = string.format( css_template, css_callback, ngx.encode_base64(json_ad) )
    ngx.say(out)
    ngx.exit(200)
end

main()

File: python/css.py
import logging, adblock, base64, json
from flask import Flask, Response, request

app = Flask(__name__)

config = {
    "logfile": "/tmp/css.log",
    "logformat": "%(asctime)-15s %(filename)s:%(lineno)-4s - %(funcName)-30s - %(message)s",
    "loglevel": logging.DEBUG,
    "logname": "css_bypass",
    "host": "0.0.0.0",
    "port": 8001,
    "css_template": '#{}{{content:"{}"}}',
    "ad_url": "http://127.0.0.1/ads/{}",
    "adblock": {
        "encode_images": True,
        "encode_videos": True
    }
}

global log
logging.basicConfig(filename=config['logfile'], level=config['loglevel'], format=config['logformat'])
log = logging.getLogger(config['logname'])

@app.route('/css/<int:ad_zone>/<css_callback>')
def handle_css(ad_zone,css_callback):
    log.debug("Received ad_zone: {} and CSS callback: {}".format( ad_zone, css_callback ))
    ad_url = config['ad_url'].format( ad_zone, request.remote_addr )
    # log.debug("Generated AD URL: {}".format(ad_url))
    (body, headers) = adblock.fetch_url( ad_url )
    # log.debug("Remote ad response body: {}".format(body))
    json_ad = adblock.process_ads_response( body, headers )
    out = config['css_template'].format( css_callback, base64.b64encode(json_ad.encode()).decode() )
    # log.debug("CSS output: {}".format(out))

    return Response(response=out, status=200, mimetype="text/css")

if __name__ == '__main__':
    log.debug("Starting up from CLI")
    app.run(host=config['host'], port=config['port'], debug=True)
else:
    log.debug("Starting up from UWSGI")
